### PR TITLE
PCA - minor changes to allow nopython jit compilation

### PR DESCRIPTION
**Pull request by rjenc29**
Small tweak to PCA to allow it to be njit compatible.  
Uses a new demean function (also njit compatible).  
Updated tests to use a wider collection of SKLearn samples and ranges of components.
Vaguely wondered if the components arg should be allowed to be None (i.e. all components) at some point, but left this for now.

### CODE REVIEW

**Comment by rjenc29** [comment]
Ha!  The docstring example is a slightly odd case as there's only in fact one PC.  Eigenvectors aren't unique and multiplying any of them by a constant (e.g. -1) gives a valid eigenvector, and vector 'sameness' is a little nuanced.  Interested in your thoughts, guys; will leave this for now.

File: fastats/scaling/scaling.py [hunk 1]
```diff
@@ -99,6 +99,22 @@ def rank(A):
...
+def demean(A):
+    """
+    Subtract the mean from the supplied data column-wise.
+    """
+    assert A.ndim > 1
+
+    n = A.shape[1]
+    res = empty_like(A, dtype=np_float64)
+
+    for i in range(n):
```
Comment (dwillmer) [inline fastats/scaling/scaling.py:111]: Does this work with `prange`?

**Comment by dwillmer** [comment]
All looks good to me, awesome stuff!

I quite like the fact that the docstring example is simple, so only being one PC is fine. The docstring is for basic API usage anyway. The rest of the tests are fairly comprehensive, so good to merge imo.

**Comment by dwillmer** [comment]
The test failure is because the PC's are agnostic to direction, so we'll need an `abs` before comparing.

**Comment by dwillmer** [comment]
Do you want to add a note to the releases about the PCA updates in this PR as well?

File: fastats/scaling/scaling.py (hunk 1 above)
Comment (dwillmer) [reply fastats/scaling/scaling.py:111]: (as in, we could have a `_parallel` variant)

Comment (rjenc29) [reply fastats/scaling/scaling.py:111]: Yep, all of the scaling functions behave well with prange (using 0.36 release candidate).  I will add parallel versions as suggested, good idea.

**Comment by rjenc29** [comment]
Sure thing, I will change the test to use abs & separately think about replicating sklearn's normalisation / deterministic sign convention.  It would however mean we hand back some of the performance gains (albeit a small amount) as it's quite intrusive.  Separate PR in due course, for discussion.

**Comment by rjenc29** [comment]
Will add a release note also

**Comment by dwillmer** [comment]
Don't worry about deterministic sign, it's not applicable for PCA and changes for each algorithm. I think an `abs` is ok for this.

**Comment by rjenc29** [comment]
Regarding the parallel business, how would you feel about a wholesale change from range to prange (well, in the places where parallelisation opportunities exist, like per-column scaling etc)?

I think it's true that if prange is encountered outside of a parallel=True JIT compilation, it just behaves like range anyway (possibly some small overhead).

Just a thought.

**Comment by dwillmer** [comment]
Wholesale change from `range` to `prange` is an interesting idea. The overhead would need to be *very* small, and it would require some changes to the ast_transforms code (`range` is a builtin so is ignored in the transforms), but that could be pretty neat.

I'd also like to know at what stage it actually performs the parallelisation - most platforms have ~100us overhead to start a thread, which we certainly don't want to add for small loops.

Maybe open a PR to throw around some ideas? Sounds like a good candidate for 2018.1.

**Comment by dwillmer** [comment]
LGTM, awesome stuff - merging.

**Pull request merged**